### PR TITLE
Improve permission handling logic

### LINKED ISSUE #730: [BUG] Internal crash when using existing external repository
**Describe the bug**
When setting an external repository, using the top bar back arrow bypasses the permission pop up dialog, which result in an internal crash.
Pressing the system back button after the crash screen makes the permission dialog appears.

Pressing the system back button instead of the top bar back arrow does not crash.

**To Reproduce**
0. Get your local git repository on your phone
1. Click on 'Use Local Directory'
2. Click on 'SD-CARD', then 'OK'
3. Select your repository
4. Click on 'Allow' (I think it is only for Android 10+)
5. Click on your back button
6. Click on the top arrow to go back (not the back button)

Using the system back arrow does bring the permission dialog.

**Expected behavior**
The storage permission pop up should appear, then the main screen should appear.

**Screenshots**
image

**Crash Log**
```
kotlin.KotlinNullPointerException
    at com.zeapo.pwdstore.utils.PasswordRepository$Companion.initialize(PasswordRepository.kt:175)
    at com.zeapo.pwdstore.PasswordStore.createRepository(PasswordStore.kt:368)
    at com.zeapo.pwdstore.PasswordStore.initializeRepositoryInfo(PasswordStore.kt:411)
    at com.zeapo.pwdstore.PasswordStore.onActivityResult(PasswordStore.kt:673)
    at android.app.Activity.dispatchActivityResult(Activity.java:8131)
    at android.app.ActivityThread.deliverResults(ActivityThread.java:4838)
    at android.app.ActivityThread.handleSendResult(ActivityThread.java:4886)
    at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
    at android.os.Handler.dispatchMessage(Handler.java:107)
    at android.os.Looper.loop(Looper.java:214)
    at android.app.ActivityThread.main(ActivityThread.java:7356)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

**Device information (please complete the following information):**
 - Device: onePlus 6T
 - OS: LineageOS 17.1
 - App version 1.7.0 / master

**Additional context**
The issue is, I think, because `onActivityResult` is called before `onResume` when using the top back arrow, so `initializeRepositoryInfo` is called before the permission dialog. 
Given that the app does not have the permission,  `dir` in `PasswordRepository` L.175 does not exist.

On the other hand, when using the system back button, no `resultCode` is set, which makes the app to bypass `onActivityResult` (`initializeRepositoryInfo` is then not called) , so `onResume` can show the permission dialog, and then everything works fine.

I think actually that the non crashing path is also wrong, because `setResult` is not called when using this button, so both path should be fixed:
- Add a callback to `onBackPressed` in order to add `setResult(Activity.RESULT_OK)` when pressing the system back button
- Make the content of `onResume` called before `onActivityResult`, or at least before `initializeRepositoryInfo`

I might have the time to make a fix for this later this week :grinning:

### PR DESCRIPTION
**Pull request by msfjarvis**
## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
Ensure storage permission is always requested when it is needed, also improve the flow around this for better UX.

## :bulb: Motivation and Context
Fixes #730

## :green_heart: How did you test it?
Copied my password store to `/sdcard/`, tried to load it using the steps in #730 and
was prompted for storage permissions rather than crashing.

## :pencil: Checklist

- [x] I ran `./gradlew spotlessApply` before submitting the PR
- [x] I reviewed submitted code

## :crystal_ball: Next steps
A lot of logic here is complex and unnecessarily tied to `UserPreference` so a refactor is in order
to move things out of it.

## :camera_flash: Screenshots / GIFs

<details>
<summary>Old permission grant flow</summary>

permission_flow_old

</details>

<details>
<summary>New permission grant flow</summary>

permission_flow

</details>

### CODE REVIEW

**Comment by msfjarvis** [comment]
I'm not very sure of the code here except that it _works_. Please feel free to make improvements.

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt [hunk 1]
```diff
@@ -413,6 +390,46 @@ class PasswordStore : AppCompatActivity() {
...
+                        .setAction(R.string.dialog_ok) {
+                            ActivityCompat.requestPermissions(
+                                    activity,
+                                    arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+                                    REQUEST_EXTERNAL_STORAGE)
+                        }
+                snack.show()
+                val view = snack.view
+                val tv: AppCompatTextView = view.findViewById(com.google.android.material.R.id.snackbar_text)
+                tv.setTextColor(Color.WHITE)
```
Comment (FabianHenneke) [inline app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:414]: Is it normal to hack on Snackbars in this way?

Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:414]: > Is it normal to hack on Snackbars in this way?

Nope. Setting the action is usually enough.

Comment (FabianHenneke) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:414]: It's also kinda weird that the snackbar is never dismissed when the completes. I will change that, but please verify that it behaves correctly.

**Comment by msfjarvis** [comment]
Seems to work as intended. The permission rationale flow actually works better in terms of UX compared to the other code path so I vote to slightly reword the Snackbar message and always use it.

**Comment by msfjarvis** [comment]
Here's before and after GIFs to compare the difference.

<details>
<summary>Old permission grant flow</summary>

permission_flow_old


</details>

<details>
<summary>New permission grant flow</summary>

permission_flow

</details>

**Comment by msfjarvis** [comment]
Gradle cache taking 4 minutes to unpack:

me:

image

**Pull request merged**